src/dotnet/CarbonAware.Tools.WattTimeClient/WattTimeClient.cs

```diff
@@ -0,0 +1,247 @@
+            if (response.IsSuccessStatusCode)
+            {
+                Log.LogDebug("Successfully retrieved {url}", uriPath);
+
+                var data = response.Content.ReadAsStringAsync();
+                return data.Result ?? string.Empty; 
+            }
+            else if (retries > 0 && retriableStatusCodes.Contains(response.StatusCode))
+            {
+                Log.LogDebug("Failed to get url {url} with status code {statusCode}.  {retries} retries remaining.", uriPath, response.StatusCode, retries);
+                retries--;
+                await this.UpdateAuthTokenAsync();
+                return await this.GetAsyncWithAuthRetries(uriPath, retries);
+            }
+            else
+            {
+                Log.LogError("Error getting data from WattTime.  StatusCode: {statusCode}", response.StatusCode);
+
+                throw new System.Exception($"Error getting data from WattTime: {response.StatusCode}");
+            }
+        }
+
+        private async Task UpdateAuthTokenAsync()
+        {
+            // Request auth token from WattTime API
+            var authToken = Encoding.ASCII.GetBytes($"{this.Configuration.Username}:{this.Configuration.Password}");
+            this.client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(Constants.BasicAuthentication, Convert.ToBase64String(authToken));
+
+            using (var activity = ActivitySource.StartActivity())
+            {
+                activity?.SetTag(Constants.Username, this.Configuration.Username);
+
+                Log.LogInformation("Attempting to log in user {username}", this.Configuration.Username);
+
+                var result = await this.client.GetStringAsync(Paths.Login);
+
+                // Store token for use with API requests
+                var data = JsonSerializer.Deserialize<LoginResult>(result, options);
+
+                if (data == null)
+                {
+                    Log.LogError("Login failed for user {username}", this.Configuration.Username);
+                    throw new AuthenticationException("Login failed.");
+                }
+
+                this.authToken = data.Token;
+                
+                this.client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue(Constants.BearerAuthentication, this.authToken);
+            }
+        }
+
+        private string BuildUrlWithQueryString(string url, IDictionary<string, string> queryStringParams)
+        {
+            if (Log.IsEnabled(LogLevel.Debug))
+            {
+                Log.LogDebug("Attempting to build a url using url {url} and query string parameters {parameters}", url, string.Join(";", queryStringParams.Select(k => $"\"{k.Key}\":\"{k.Value}\"")));
+            }
+
+            var query = HttpUtility.ParseQueryString("");
+            foreach(var kvp in queryStringParams)
+            {
+                query[kvp.Key] = kvp.Value;
+            }
+
+            var result = $"{url}?{query}";
```

Comment:
I'm assuming that ParseQueryString() doest the  `UrlEncode` for the query string for you? Otherwise it has to be done.